### PR TITLE
Call PlayerReadyArrowEvent for when items in the offhand are used

### LINKED ISSUE #8841: PlayerReadyArrowEvent does not fire when players hold the arrow in their offhand
### Expected behavior

That the event PlayerReadyArrowEvent fires when the player holds an arrow in their offhand, which is prioritised over other arrows in the inventory.

### Observed/Actual behavior

The event, well, does not fire.

### Steps/models to reproduce

1. Have a plugin listening to the event.
2. Hold an arrow in the hotbar, confirm that it fires.
3. Hold an arrow in your offhand, confirm that it doesn't fire despite the arrow being used when firing.

### Plugin and Datapack List

All plugins for a test server and confirmed elsewhere but just to confirm:
- AdvancedTeleport
- ATHENA
- ClearLag
- CratesPlus
- HeadsPlus
- LuckPerms
- Multiverse-Core
- PlaceholderAPI
- Spoofer
- Vault
- ViaVersion

Datapacks:
- vanilla
- file/bukkit

### Paper version

This server is running Paper version git-Paper-396 (MC: 1.19.3) (Implementing API version 1.19.3-R0.1-SNAPSHOT) (Git: bd77b78)
You are running the latest version
Previous version: git-Paper-379 (MC: 1.19.3)

### Other

_No response_

### PR DESCRIPTION
**Pull request by Thatsmusic99**
Fixes #8841 

Combines the existing item predicates with the tryReadyArrow method.

### CODE REVIEW

File: patches/server/0210-PlayerReadyArrowEvent.patch [hunk 1]
```diff
@@ -28,12 +28,22 @@ index e088e5174a4666130cd1f66c751e7ab83d9f7a65..8aa5f8bacb868a3b04af6e696f9eed89
      @Override
      public ItemStack getProjectile(ItemStack stack) {
          if (!(stack.getItem() instanceof ProjectileWeaponItem)) {
-@@ -2242,7 +2253,7 @@ public abstract class Player extends LivingEntity {
+             return ItemStack.EMPTY;
+         } else {
+-            Predicate<ItemStack> predicate = ((ProjectileWeaponItem) stack.getItem()).getSupportedHeldProjectiles();
++            Predicate<ItemStack> predicate = (item) -> ((ProjectileWeaponItem) stack.getItem()).getSupportedHeldProjectiles().test(item) && tryReadyArrow(stack, item); // Paper
```
Comment (Machine-Maker) [inline patches/server/0210-PlayerReadyArrowEvent.patch]: You should combine these predicates with the instance `and(Predicate)` method instead.

File: patches/server/0210-PlayerReadyArrowEvent.patch [hunk 2]
```diff
@@ -28,12 +28,22 @@ index e088e5174a4666130cd1f66c751e7ab83d9f7a65..8aa5f8bacb868a3b04af6e696f9eed89
...
+             ItemStack itemstack1 = ProjectileWeaponItem.getHeldProjectile(this, predicate);
+ 
+             if (!itemstack1.isEmpty()) {
+                 return itemstack1;
+             } else {
+-                predicate = ((ProjectileWeaponItem) stack.getItem()).getAllSupportedProjectiles();
++                predicate = (item) -> ((ProjectileWeaponItem) stack.getItem()).getAllSupportedProjectiles().test(item) && tryReadyArrow(stack, item); // Paper
```
Comment (Machine-Maker) [inline patches/server/0210-PlayerReadyArrowEvent.patch]: see above

**Review by Machine-Maker — CHANGES_REQUESTED** [review]
Looks good, just 1 comment, and I think you added some other random commits to the branch, like "Added exception printing when player data fails to load".

**Comment by Thatsmusic99** [comment]
Crap, thought I avoided that, I had my fork originally used for debugging. Let me review that.

File: patches/server/0210-PlayerReadyArrowEvent.patch (hunk 1 above)
Comment (Thatsmusic99) [reply patches/server/0210-PlayerReadyArrowEvent.patch]: Updated using this method.

**Comment by Thatsmusic99** [comment]
Should've removed the erroneous patch now, did notice the hashes in some of the other hash files changing, not sure how big of a deal that is.

File: patches/server/0210-PlayerReadyArrowEvent.patch [hunk 3]
```diff
@@ -28,12 +28,23 @@ index e088e5174a4666130cd1f66c751e7ab83d9f7a65..8aa5f8bacb868a3b04af6e696f9eed89
...
+             } else {
+-                predicate = ((ProjectileWeaponItem) stack.getItem()).getAllSupportedProjectiles();
++                predicate = ((ProjectileWeaponItem) stack.getItem()).getAllSupportedProjectiles().and(item -> tryReadyArrow(stack, item)); // Paper
+ 
                  for (int i = 0; i < this.inventory.getContainerSize(); ++i) {
                      ItemStack itemstack2 = this.inventory.getItem(i);
  
 -                    if (predicate.test(itemstack2)) {
-+                    if (predicate.test(itemstack2) && tryReadyArrow(stack, itemstack2)) { // Paper
++                    if (predicate.test(itemstack2)) {
```
Comment (Machine-Maker) [inline patches/server/0210-PlayerReadyArrowEvent.patch]: Looks like you added a space at the end of this line causing it to show up in the diff when it should be an "unchanged" line as you removed the diff that was originally added.

**Comment by Machine-Maker** [comment]
> did notice the hashes in some of the other hash files changing, not sure how big of a deal that is.

nah, that's fine. can just ignore that.

File: patches/server/0210-PlayerReadyArrowEvent.patch (hunk 3 above)
Comment (Thatsmusic99) [reply patches/server/0210-PlayerReadyArrowEvent.patch]: Removed now, came from the paper comment that was originally there.

**Review by Machine-Maker — APPROVED** [review]

**Comment by Doc94** [comment]
maybe im confuse this is not related to https://github.com/PaperMC/Paper/pull/7714 ? i feel when make that PR test this... or was for the not-bow things...

**Comment by Machine-Maker** [comment]
Well, in that PR, you don't check each hand individually, I think its better do do that. That PR just checks the result of checking both hands. Also this PR handles it in a much cleaner way, just adding the check into the predicate once. I also don't there there's really a need for a separate event just to fix naming issues. Or if there is, it should be handled separately. Also it looks like there are some other issues with that PR.

**Comment by Doc94** [comment]
> Well, in that PR, you don't check each hand individually, I think its better do do that. That PR just checks the result of checking both hands. Also this PR handles it in a much cleaner way, just adding the check into the predicate once. I also don't there there's really a need for a separate event just to fix naming issues. Or if there is, it should be handled separately. Also it looks like there are some other issues with that PR.

okay, originally that PR was only for try fix this and also include crossbow with fireworks, but get the recommendation for make a new event for avoid issues with the name.. also yes this PR handle better the thing about check the offhand... then maybe can close my original PR and wait for the merge of this to make a new PR for only the thing about the crossbows/fireworks?

**Comment by Thatsmusic99** [comment]
I have tested and confirmed that with this PR the event fires with crossbows, and when using fireworks. The only potential issue I see with the current fix is that it doesn't fire when preparing an arrow in creative mode because, well, said item would be null. Have no clue whether that is intentional or not.

**Comment by lynxplay** [comment]
That was the case before as well, generally there isn't a consistent arrow item that the player is preparing so I don't think we necessarily have to fire and event here :+1:

**Review by lynxplay — APPROVED** [review]

**Review by kennytv — APPROVED** [review]

**Pull request merged**